python/lsst/ctrl/oods/gen3ButlerBroker.py

```diff
@@ -0,0 +1,260 @@
+
+        Returns
+        -------
+        info: `dict`
+            Dictionary containing file name, and uninitialized elements
+        """
+        info = dict()
+        info['FILENAME'] = os.path.basename(filename)
+        info['CAMERA'] = ''
+        info['OBSID'] = ''
+        info['RAFT'] = ''
+        info['SENSOR'] = ''
+        return info
+
+    def transmit_status(self, metadata, code, description):
+        """Transmit a message with given metadata, status code and description
+
+        Parameters
+        ----------
+        metadata: `dict`
+            dictionary containing meta data about the image
+        code: `int`
+            status code
+        description: `str`
+            description of the ingestion status
+        """
+        msg = dict(metadata)
+        msg['MSG_TYPE'] = 'IMAGE_IN_OODS'
+        msg['ARCHIVER'] = self.archiver_name
+        msg['STATUS_CODE'] = code
+        msg['DESCRIPTION'] = description
+        LOGGER.info(f"msg: {msg}, code: {code}, description: {description}")
+        if self.publisher is None:
+            return
+        asyncio.create_task(self.publisher.publish_message(self.publisher_queue, msg))
+
+    def on_success(self, datasets):
+        """Callback used on successful ingest. Used to transmit
+        successful data ingestion status
+
+        Parameters
+        ----------
+        datasets: `list`
+            list of DatasetRefs
+        """
+        for dataset in datasets:
+            LOGGER.info(f"{dataset.path.ospath} file ingested")
+            image_data = ImageData(dataset)
+            self.transmit_status(image_data.info, code=0, description="file ingested")
+
+    def on_ingest_failure(self, filename, exc):
+        """Callback used on ingest failure. Used to transmit
+        unsuccessful data ingestion status
+
+        Parameters
+        ----------
+        filename: `ButlerURI`
+            ButlerURI that failed in ingest
+        exc: `Exception`
+            Exception which explains what happened
+
+        """
+        real_file = filename.ospath
+        self.move_file_to_bad_dir(real_file)
+        cause = self.extract_cause(exc)
+        info = self.undef_metadata(real_file)
+        self.transmit_status(info, code=1, description=f"ingest failure: {cause}")
+
+    def on_metadata_failure(self, filename, exc):
+        """Callback used on metadata extraction failure. Used to transmit
+        unsuccessful data ingestion status
+
+        Parameters
+        ----------
+        filename: `ButlerURI`
+            ButlerURI that failed in ingest
+        exc: `Exception`
+            Exception which explains what happened
+        """
+        real_file = filename.ospath
+        self.move_file_to_bad_dir(real_file)
+
+        cause = self.extract_cause(exc)
+        info = self.undef_metadata(real_file)
+        self.transmit_status(info, code=2, description=f"metadata failure: {cause}")
+
+    def move_file_to_bad_dir(self, filename):
```

Comment:
The docstring is missing (if the method is considered to be public).

Reply:
Added